File: android/versioned-abis/expoview-abi42_0_0/src/main/java/abi42_0_0/host/exp/exponent/modules/api/screens/ScreenWindowTraits.java
package abi42_0_0.host.exp.exponent.modules.api.screens;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.pm.ActivityInfo;
import android.os.Build;
import android.view.View;
import android.view.ViewParent;
import android.view.WindowInsets;
import android.view.WindowManager;

import androidx.annotation.Nullable;
import androidx.core.view.ViewCompat;

import abi42_0_0.com.facebook.react.bridge.GuardedRunnable;
import abi42_0_0.com.facebook.react.bridge.ReactContext;
import abi42_0_0.com.facebook.react.bridge.UiThreadUtil;

public class ScreenWindowTraits {
  // Methods concerning statusBar management were taken from `react-native`'s status bar module:
  // https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/statusbar/StatusBarModule.java

  private static boolean mDidSetOrientation = false;
  private static boolean mDidSetStatusBarAppearance = false;
  private static Integer mDefaultStatusBarColor = null;

  protected static void applyDidSetOrientation() {
    mDidSetOrientation = true;
  }

  public static boolean didSetOrientation() {
    return mDidSetOrientation;
  }

  protected static void applyDidSetStatusBarAppearance() {
    mDidSetStatusBarAppearance = true;
  }

  public static boolean didSetStatusBarAppearance() {
    return mDidSetStatusBarAppearance;
  }

  protected static void setOrientation(Screen screen, final Activity activity) {
    if (activity == null) {
      return;
    }

    Screen screenForOrientation =  ScreenWindowTraits.findScreenForTrait(screen, Screen.WindowTraits.ORIENTATION);

    final Integer orientation;

    if (screenForOrientation != null && screenForOrientation.getScreenOrientation() != null) {
      orientation = screenForOrientation.getScreenOrientation();
    } else {
      orientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED;
    }

    activity.setRequestedOrientation(orientation);
  }

  protected static void setColor(Screen screen, final Activity activity, ReactContext context) {
    if (activity == null || context == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      return;
    }

    if (mDefaultStatusBarColor == null) {
      mDefaultStatusBarColor = activity.getWindow().getStatusBarColor();
    }

    Screen screenForColor = ScreenWindowTraits.findScreenForTrait(screen, Screen.WindowTraits.COLOR);
    Screen screenForAnimated = ScreenWindowTraits.findScreenForTrait(screen, Screen.WindowTraits.ANIMATED);

    final Integer color;
    final boolean animated;

    if (screenForColor != null && screenForColor.getStatusBarColor() != null) {
      color = screenForColor.getStatusBarColor();
    } else {
      color = mDefaultStatusBarColor;
    }

    if (screenForAnimated != null && screenForAnimated.isStatusBarAnimated() != null) {
      animated = screenForAnimated.isStatusBarAnimated();
    } else {
      animated = false;
    }

    UiThreadUtil.runOnUiThread(
            new GuardedRunnable(context) {
              @TargetApi(Build.VERSION_CODES.LOLLIPOP)
              @Override
              public void runGuarded() {
                activity
                        .getWindow()
                        .addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
                int curColor = activity.getWindow().getStatusBarColor();
                ValueAnimator colorAnimation =
                        ValueAnimator.ofObject(new ArgbEvaluator(), curColor, color);
                colorAnimation.addUpdateListener(
                        new ValueAnimator.AnimatorUpdateListener() {
                          @Override
                          public void onAnimationUpdate(ValueAnimator animator) {
                            activity.getWindow().setStatusBarColor((Integer) animator.getAnimatedValue());
                          }
                        });

                if (animated) {
                  colorAnimation.setDuration(300).setStartDelay(0);
                } else {
                  colorAnimation.setDuration(0).setStartDelay(300);
                }
                colorAnimation.start();
              }
            });
  }

  protected static void setStyle(Screen screen, final Activity activity, ReactContext context) {
    if (activity == null || context == null) {
      return;
    }

    Screen screenForStyle =  ScreenWindowTraits.findScreenForTrait(screen, Screen.WindowTraits.STYLE);

    final String style;

    if (screenForStyle != null && screenForStyle.getStatusBarStyle() != null) {
      style = screenForStyle.getStatusBarStyle();
    } else {
      style = "light";
    }

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      UiThreadUtil.runOnUiThread(
              new Runnable() {
                @TargetApi(Build.VERSION_CODES.M)
                @Override
                public void run() {
                  View decorView = activity.getWindow().getDecorView();
                  int systemUiVisibilityFlags = decorView.getSystemUiVisibility();
                  if ("dark".equals(style)) {
                    systemUiVisibilityFlags |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
                  } else {
                    systemUiVisibilityFlags &= ~View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
                  }
                  decorView.setSystemUiVisibility(systemUiVisibilityFlags);
                }
              });
    }
  }

  protected static void setTranslucent(Screen screen, final Activity activity, ReactContext context) {
    if (activity == null || context == null) {
      return;
    }

    final boolean translucent;

    Screen screenForTranslucent = ScreenWindowTraits.findScreenForTrait(screen, Screen.WindowTraits.TRANSLUCENT);

    if (screenForTranslucent != null && screenForTranslucent.isStatusBarTranslucent() != null) {
      translucent = screenForTranslucent.isStatusBarTranslucent();
    } else {
      translucent = false;
    }

    UiThreadUtil.runOnUiThread(
            new GuardedRunnable(context) {
              @TargetApi(Build.VERSION_CODES.LOLLIPOP)
              @Override
              public void runGuarded() {
                // If the status bar is translucent hook into the window insets calculations
                // and consume all the top insets so no padding will be added under the status bar.
                View decorView = activity.getWindow().getDecorView();
                if (translucent) {
                  decorView.setOnApplyWindowInsetsListener(
                          new View.OnApplyWindowInsetsListener() {
                            @Override
                            public WindowInsets onApplyWindowInsets(View v, WindowInsets insets) {
                              WindowInsets defaultInsets = v.onApplyWindowInsets(insets);
                              return defaultInsets.replaceSystemWindowInsets(
                                      defaultInsets.getSystemWindowInsetLeft(),
                                      0,
                                      defaultInsets.getSystemWindowInsetRight(),
                                      defaultInsets.getSystemWindowInsetBottom());
                            }
                          });
                } else {
                  decorView.setOnApplyWindowInsetsListener(null);
                }

                ViewCompat.requestApplyInsets(decorView);
              }
            });
  }

  protected static void setHidden(Screen screen, final Activity activity) {
    if (activity == null) {
      return;
    }

    final boolean hidden;

    Screen screenForHidden = ScreenWindowTraits.findScreenForTrait(screen, Screen.WindowTraits.HIDDEN);

    if (screenForHidden != null && screenForHidden.isStatusBarHidden() != null) {
      hidden = screenForHidden.isStatusBarHidden();
    } else {
      hidden = false;
    }

    UiThreadUtil.runOnUiThread(
            new Runnable() {
              @Override
              public void run() {
                if (hidden) {
                  activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                  activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                } else {
                  activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                  activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
                }
              }
            });
  }

  protected static void trySetWindowTraits(Screen screen, Activity activity, ReactContext context) {
    if (ScreenWindowTraits.didSetOrientation()) {
      ScreenWindowTraits.setOrientation(screen, activity);
    }
    if (ScreenWindowTraits.didSetStatusBarAppearance()) {
      ScreenWindowTraits.setColor(screen, activity, context);
      ScreenWindowTraits.setStyle(screen, activity, context);
      ScreenWindowTraits.setTranslucent(screen, activity, context);
      ScreenWindowTraits.setHidden(screen, activity);
    }
  }

  protected static Screen findScreenForTrait(Screen screen, Screen.WindowTraits trait) {
    Screen childWithTrait = childScreenWithTraitSet(screen, trait);
    if (childWithTrait != null) {
      return childWithTrait;
    }

    if (checkTraitForScreen(screen, trait)) {
      return screen;
    } else {
      // if there is no child with trait set and this screen has no trait set, we look for a parent that has the trait set
      return findParentWithTraitSet(screen, trait);
    }
  }

  private static @Nullable Screen findParentWithTraitSet(Screen screen, Screen.WindowTraits trait) {
    ViewParent parent = screen.getContainer();
    while (parent != null) {
      if (parent instanceof Screen) {
        if (checkTraitForScreen((Screen) parent, trait)) {
          return (Screen) parent;
        }
      }
      parent = parent.getParent();
    }
    return null;
  }

  protected static @Nullable Screen childScreenWithTraitSet(Screen screen, Screen.WindowTraits trait) {
    if (screen == null || screen.getFragment() == null) {
      return null;
    }
    for (ScreenContainer sc : screen.getFragment().getChildScreenContainers()) {
      // we check only the top screen for the trait
      Screen topScreen = sc.getTopScreen();
      Screen child = childScreenWithTraitSet(topScreen, trait);
      if (child != null) {
        return child;
      }
      if (topScreen != null && checkTraitForScreen(topScreen, trait)) {
        return topScreen;
      }
    }
    return null;
  }

  private static boolean checkTraitForScreen(Screen screen, Screen.WindowTraits trait) {
    switch (trait) {
      case ORIENTATION:
        return screen.getScreenOrientation() != null;
      case COLOR:
        return screen.getStatusBarColor() != null;
      case STYLE:
        return screen.getStatusBarStyle() != null;
      case TRANSLUCENT:
        return screen.isStatusBarTranslucent() != null;
      case HIDDEN:
        return screen.isStatusBarHidden() != null;
      case ANIMATED:
        return screen.isStatusBarAnimated() != null;
      default:
        throw new IllegalArgumentException("Wrong trait passed: " + trait);
    }
  }

}

File: android/versioned-abis/expoview-abi44_0_0/src/main/java/abi44_0_0/host/exp/exponent/modules/api/screens/ScreenWindowTraits.kt
package abi44_0_0.host.exp.exponent.modules.api.screens

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.app.Activity
import android.content.pm.ActivityInfo
import android.os.Build
import android.view.View
import android.view.ViewParent
import android.view.WindowManager
import androidx.core.view.ViewCompat
import abi44_0_0.com.facebook.react.bridge.GuardedRunnable
import abi44_0_0.com.facebook.react.bridge.ReactContext
import abi44_0_0.com.facebook.react.bridge.UiThreadUtil
import abi44_0_0.host.exp.exponent.modules.api.screens.Screen.WindowTraits

object ScreenWindowTraits {
  // Methods concerning statusBar management were taken from `react-native`'s status bar module:
  // https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/statusbar/StatusBarModule.java
  private var mDidSetOrientation = false
  private var mDidSetStatusBarAppearance = false
  private var mDefaultStatusBarColor: Int? = null
  internal fun applyDidSetOrientation() {
    mDidSetOrientation = true
  }

  internal fun applyDidSetStatusBarAppearance() {
    mDidSetStatusBarAppearance = true
  }

  internal fun setOrientation(screen: Screen, activity: Activity?) {
    if (activity == null) {
      return
    }
    val screenForOrientation = findScreenForTrait(screen, WindowTraits.ORIENTATION)
    val orientation = screenForOrientation?.screenOrientation ?: ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
    activity.requestedOrientation = orientation
  }

  @SuppressLint("ObsoleteSdkInt") // to be removed when support for < 0.64 is dropped
  internal fun setColor(screen: Screen, activity: Activity?, context: ReactContext?) {
    if (activity == null || context == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      return
    }
    if (mDefaultStatusBarColor == null) {
      mDefaultStatusBarColor = activity.window.statusBarColor
    }
    val screenForColor = findScreenForTrait(screen, WindowTraits.COLOR)
    val screenForAnimated = findScreenForTrait(screen, WindowTraits.ANIMATED)
    val color = screenForColor?.statusBarColor ?: mDefaultStatusBarColor
    val animated = screenForAnimated?.isStatusBarAnimated ?: false

    UiThreadUtil.runOnUiThread(
      object : GuardedRunnable(context) {
        override fun runGuarded() {
          activity
            .window
            .addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
          val curColor = activity.window.statusBarColor
          val colorAnimation = ValueAnimator.ofObject(ArgbEvaluator(), curColor, color)
          colorAnimation.addUpdateListener { animator -> activity.window.statusBarColor = (animator.animatedValue as Int) }
          if (animated) {
            colorAnimation.setDuration(300).startDelay = 0
          } else {
            colorAnimation.setDuration(0).startDelay = 300
          }
          colorAnimation.start()
        }
      })
  }

  internal fun setStyle(screen: Screen, activity: Activity?, context: ReactContext?) {
    if (activity == null || context == null) {
      return
    }
    val screenForStyle = findScreenForTrait(screen, WindowTraits.STYLE)
    val style = screenForStyle?.statusBarStyle ?: "light"

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      UiThreadUtil.runOnUiThread {
        val decorView = activity.window.decorView
        var systemUiVisibilityFlags = decorView.systemUiVisibility
        systemUiVisibilityFlags = if ("dark" == style) {
          systemUiVisibilityFlags or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
        } else {
          systemUiVisibilityFlags and View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR.inv()
        }
        decorView.systemUiVisibility = systemUiVisibilityFlags
      }
    }
  }

  internal fun setTranslucent(
    screen: Screen,
    activity: Activity?,
    context: ReactContext?
  ) {
    if (activity == null || context == null) {
      return
    }
    val screenForTranslucent = findScreenForTrait(screen, WindowTraits.TRANSLUCENT)
    val translucent = screenForTranslucent?.isStatusBarTranslucent ?: false
    UiThreadUtil.runOnUiThread(
      object : GuardedRunnable(context) {
        @TargetApi(Build.VERSION_CODES.LOLLIPOP)
        override fun runGuarded() {
          // If the status bar is translucent hook into the window insets calculations
          // and consume all the top insets so no padding will be added under the status bar.
          val decorView = activity.window.decorView
          if (translucent) {
            decorView.setOnApplyWindowInsetsListener { v, insets ->
              val defaultInsets = v.onApplyWindowInsets(insets)
              defaultInsets.replaceSystemWindowInsets(
                defaultInsets.systemWindowInsetLeft,
                0,
                defaultInsets.systemWindowInsetRight,
                defaultInsets.systemWindowInsetBottom
              )
            }
          } else {
            decorView.setOnApplyWindowInsetsListener(null)
          }
          ViewCompat.requestApplyInsets(decorView)
        }
      })
  }

  internal fun setHidden(screen: Screen, activity: Activity?) {
    if (activity == null) {
      return
    }
    val screenForHidden = findScreenForTrait(screen, WindowTraits.HIDDEN)
    val hidden = screenForHidden?.isStatusBarHidden ?: false
    UiThreadUtil.runOnUiThread {
      if (hidden) {
        activity.window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
        activity.window.clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN)
      } else {
        activity.window.addFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN)
        activity.window.clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
      }
    }
  }

  internal fun trySetWindowTraits(screen: Screen, activity: Activity?, context: ReactContext?) {
    if (mDidSetOrientation) {
      setOrientation(screen, activity)
    }
    if (mDidSetStatusBarAppearance) {
      setColor(screen, activity, context)
      setStyle(screen, activity, context)
      setTranslucent(screen, activity, context)
      setHidden(screen, activity)
    }
  }

  private fun findScreenForTrait(screen: Screen, trait: WindowTraits): Screen? {
    val childWithTrait = childScreenWithTraitSet(screen, trait)
    if (childWithTrait != null) {
      return childWithTrait
    }
    return if (checkTraitForScreen(screen, trait)) {
      screen
    } else {
      // if there is no child with trait set and this screen has no trait set, we look for a parent
      // that has the trait set
      findParentWithTraitSet(screen, trait)
    }
  }

  private fun findParentWithTraitSet(screen: Screen, trait: WindowTraits): Screen? {
    var parent: ViewParent? = screen.container
    while (parent != null) {
      if (parent is Screen) {
        if (checkTraitForScreen(parent, trait)) {
          return parent
        }
      }
      parent = parent.parent
    }
    return null
  }

  private fun childScreenWithTraitSet(
    screen: Screen?,
    trait: WindowTraits
  ): Screen? {
    screen?.fragment?.let {
      for (sc in it.childScreenContainers) {
        // we check only the top screen for the trait
        val topScreen = sc.topScreen
        val child = childScreenWithTraitSet(topScreen, trait)
        if (child != null) {
          return child
        }
        if (topScreen != null && checkTraitForScreen(topScreen, trait)) {
          return topScreen
        }
      }
    }
    return null
  }

  private fun checkTraitForScreen(screen: Screen, trait: WindowTraits): Boolean {
    return when (trait) {
      WindowTraits.ORIENTATION -> screen.screenOrientation != null
      WindowTraits.COLOR -> screen.statusBarColor != null
      WindowTraits.STYLE -> screen.statusBarStyle != null
      WindowTraits.TRANSLUCENT -> screen.isStatusBarTranslucent != null
      WindowTraits.HIDDEN -> screen.isStatusBarHidden != null
      WindowTraits.ANIMATED -> screen.isStatusBarAnimated != null
    }
  }
}
